File: backend/app/api/v1/uploads.py
"""Upload routes — Owner: B1."""

from __future__ import annotations

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Body, Depends, File, Form, UploadFile, status
from fastapi.responses import FileResponse, Response

from app.core.deps import DbSession, get_current_user
from app.models.user import User
from app.schemas.upload import (
    ClientPlatform,
    CompleteUploadSessionRequest,
    CompleteUploadSessionResponse,
    CreateUploadSessionRequest,
    UploadOut,
    UploadPurpose,
    UploadSessionOut,
)
from app.services.event_service import event_service
from app.services.resumable_upload_service import resumable_upload_service
from app.services.upload_service import upload_service

router = APIRouter()


@router.post("", response_model=UploadOut, status_code=status.HTTP_201_CREATED)
async def create_upload(
    db: DbSession,
    user: Annotated[User, Depends(get_current_user)],
    file: Annotated[UploadFile, File()],
    purpose: Annotated[UploadPurpose, Form()] = "message_attachment",
    conversation_id: Annotated[UUID | None, Form()] = None,
    client_platform: Annotated[ClientPlatform, Form()] = "web",
) -> UploadOut:
    if conversation_id is not None:
        await upload_service.assert_conversation_owner(
            db,
            user_id=user.id,
            conversation_id=conversation_id,
        )
    upload = await upload_service.create_upload(
        db,
        user_id=user.id,
        conversation_id=conversation_id,
        purpose=purpose,
        file=file,
        client_platform=client_platform,
    )
    await event_service.record(
        db,
        user_id=user.id,
        event_type="upload.ready",
        resource_type="upload",
        resource_id=upload.id,
        conversation_id=upload.conversation_id,
        payload={"client_platform": upload.client_platform},
    )
    await db.commit()
    await db.refresh(upload)
    return upload_service.to_out(upload)


@router.post("/sessions", response_model=UploadSessionOut, status_code=status.HTTP_201_CREATED)
async def create_upload_session(
    payload: CreateUploadSessionRequest,
    db: DbSession,
    user: Annotated[User, Depends(get_current_user)],
) -> UploadSessionOut:
    if payload.conversation_id is not None:
        await upload_service.assert_conversation_owner(
            db,
            user_id=user.id,
            conversation_id=payload.conversation_id,
        )
    session = await resumable_upload_service.create(db, user_id=user.id, payload=payload)
    await db.commit()
    await db.refresh(session)
    return resumable_upload_service.to_out(session)


@router.get("/sessions/{session_id}", response_model=UploadSessionOut)
async def get_upload_session(
    session_id: UUID,
    db: DbSession,
    user: Annotated[User, Depends(get_current_user)],
) -> UploadSessionOut:
    session = await resumable_upload_service.get_owned(
        db,
        user_id=user.id,
        session_id=session_id,
    )
    return resumable_upload_service.to_out(session)


@router.put("/sessions/{session_id}/parts/{part_number}", response_model=UploadSessionOut)
async def put_upload_session_part(
    session_id: UUID,
    part_number: int,
    db: DbSession,
    user: Annotated[User, Depends(get_current_user)],
    body: Annotated[bytes, Body(media_type="application/octet-stream")],
) -> UploadSessionOut:
    session = await resumable_upload_service.put_part(
        db,
        user_id=user.id,
        session_id=session_id,
        part_number=part_number,
        data=body,
    )
    await db.commit()
    await db.refresh(session)
    return resumable_upload_service.to_out(session)


@router.post("/sessions/{session_id}/complete")
async def complete_upload_session(
    session_id: UUID,
    payload: CompleteUploadSessionRequest,
    db: DbSession,
    user: Annotated[User, Depends(get_current_user)],
) -> CompleteUploadSessionResponse:
    session, upload = await resumable_upload_service.complete(
        db,
        user_id=user.id,
        session_id=session_id,
        payload=payload,
    )
    await db.commit()
    await db.refresh(session)
    return CompleteUploadSessionResponse(
        session=resumable_upload_service.to_out(session),
        upload=upload,
    )


@router.delete("/sessions/{session_id}", response_model=UploadSessionOut)
async def cancel_upload_session(
    session_id: UUID,
    db: DbSession,
    user: Annotated[User, Depends(get_current_user)],
) -> UploadSessionOut:
    session = await resumable_upload_service.cancel(db, user_id=user.id, session_id=session_id)
    await db.commit()
    await db.refresh(session)
    return resumable_upload_service.to_out(session)


@router.get("/{upload_id}", response_model=UploadOut)
async def get_upload(
    upload_id: UUID,
    db: DbSession,
    user: Annotated[User, Depends(get_current_user)],
) -> UploadOut:
    upload = await upload_service.get_owned_upload(db, user_id=user.id, upload_id=upload_id)
    return upload_service.to_out(upload)


@router.get("/{upload_id}/download")
async def download_upload(
    upload_id: UUID,
    db: DbSession,
    user: Annotated[User, Depends(get_current_user)],
) -> FileResponse:
    upload = await upload_service.get_owned_upload(db, user_id=user.id, upload_id=upload_id)
    path = upload_service.download_path(upload)
    return FileResponse(
        path,
        filename=upload.filename,
        media_type=upload.detected_content_type or upload.content_type,
        headers={"X-Content-Type-Options": "nosniff"},
    )


@router.delete("/{upload_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_upload(
    upload_id: UUID,
    db: DbSession,
    user: Annotated[User, Depends(get_current_user)],
) -> Response:
    upload = await upload_service.get_owned_upload(db, user_id=user.id, upload_id=upload_id)
    await upload_service.delete_upload(db, upload)
    await db.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/lib/adapters/uploads.ts
import { api } from '@/lib/api';
import type { UploadOut, UploadPurpose } from '@/lib/types';

export interface UploadSessionOut {
  id: string;
  filename: string;
  content_type: string;
  total_size_bytes: number;
  expected_sha256?: string | null;
  client_platform: 'web' | 'desktop' | 'ios' | 'android';
  part_size_bytes: number;
  received_parts: number[];
  status: 'open' | 'completed' | 'cancelled' | 'failed' | 'expired';
  upload_id?: string | null;
  error_code?: string | null;
  error_message?: string | null;
  created_at: string;
  updated_at: string;
  expires_at: string;
}

export interface CreateUploadSessionInput {
  filename: string;
  contentType: string;
  totalSizeBytes: number;
  purpose: UploadPurpose;
  conversationId?: string;
  expectedSha256?: string;
  clientPlatform?: 'web' | 'ios' | 'android' | 'desktop';
  partSizeBytes?: number;
}

export interface UploadFileInput {
  file: File | Blob;
  filename: string;
  purpose: UploadPurpose;
  conversationId?: string;
  clientPlatform?: 'web' | 'ios' | 'android' | 'desktop';
  signal?: AbortSignal;
  onProgress?: (progress: number) => void;
}

export async function uploadFile(input: UploadFileInput): Promise<UploadOut> {
  const formData = new FormData();
  formData.append('file', input.file, input.filename);
  formData.append('purpose', input.purpose);
  if (input.conversationId) formData.append('conversation_id', input.conversationId);
  if (input.clientPlatform) formData.append('client_platform', input.clientPlatform);

  const { data } = await api.post<UploadOut>('/api/v1/uploads', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
    signal: input.signal,
    onUploadProgress: (event) => {
      if (!input.onProgress || !event.total) return;
      input.onProgress(Math.round((event.loaded / event.total) * 100));
    },
  });
  return data;
}

export async function deleteUpload(uploadId: string): Promise<void> {
  await api.delete(`/api/v1/uploads/${uploadId}`);
}

export async function downloadUpload(uploadId: string): Promise<Blob> {
  const { data } = await api.get<Blob>(`/api/v1/uploads/${uploadId}/download`, {
    responseType: 'blob',
  });
  return data;
}

export async function createUploadSession(input: CreateUploadSessionInput): Promise<UploadSessionOut> {
  const { data } = await api.post<UploadSessionOut>('/api/v1/uploads/sessions', {
    filename: input.filename,
    content_type: input.contentType,
    total_size_bytes: input.totalSizeBytes,
    purpose: input.purpose,
    conversation_id: input.conversationId,
    expected_sha256: input.expectedSha256,
    client_platform: input.clientPlatform ?? 'web',
    part_size_bytes: input.partSizeBytes,
  });
  return data;
}

export async function getUploadSession(sessionId: string): Promise<UploadSessionOut> {
  const { data } = await api.get<UploadSessionOut>(`/api/v1/uploads/sessions/${sessionId}`);
  return data;
}

export async function putUploadSessionPart(
  sessionId: string,
  partNumber: number,
  data: Blob | ArrayBuffer,
): Promise<UploadSessionOut> {
  const { data: session } = await api.put<UploadSessionOut>(
    `/api/v1/uploads/sessions/${sessionId}/parts/${partNumber}`,
    data,
    { headers: { 'Content-Type': 'application/octet-stream' } },
  );
  return session;
}

export async function completeUploadSession(
  sessionId: string,
  sha256?: string,
): Promise<{ session: UploadSessionOut; upload: UploadOut }> {
  const { data } = await api.post<{ session: UploadSessionOut; upload: UploadOut }>(
    `/api/v1/uploads/sessions/${sessionId}/complete`,
    { sha256 },
  );
  return data;
}

export async function cancelUploadSession(sessionId: string): Promise<UploadSessionOut> {
  const { data } = await api.delete<UploadSessionOut>(`/api/v1/uploads/sessions/${sessionId}`);
  return data;
}

export function uploadDownloadUrl(uploadId: string, variant: 'original' | 'thumbnail' = 'original') {
  const query = variant === 'original' ? '' : `?variant=${variant}`;
  return `/api/v1/uploads/${uploadId}/download${query}`;
}
